Add initial BM Flow docs (#2290)
wix/yoshi: website/docs/business-manager-flow.md

---
id: business-manager-flow

> title: Business-Manager Flow Overview

Reviewer: Is this an overview?
```suggestion
title: Business-Manager Flow
```

I think it's the whole docs 🥼 

(BTW I know that editor-flow has the same @yavorsky I think we should modify that as well)

sidebar_label: Business-Manager Flow
---

> 🧪 Experimental

## Overview

The Business-Manager is Wix's backoffice platform.
For more information on the platform, see the [official docs](https://github.com/wix-private/business-manager/).

> Yoshi's Business-Manager Flow favors Convention over Configuration by abstracting away the following concepts:

Reviewer: ```suggestion
Yoshi's Business-Manager Flow ease the development process of business-manager modules by reducing boilerplate, automating common tasks and solving infrastructure and runtime problems in a single place.

Inspired by [next.js](https://nextjs.org/#file-system-routing) It requires the developer to create a specific file structure and can statically analyze it to provide features like file system routing, auto generation of module config and more.
```

> 

Reviewer: This section needs a header
```suggestion
## Concepts
```

For example:

| Path                    | Route      |
| ----------------------- | ---------- |
| `src/pages/index.tsx`   | `/`        |
| `src/pages/foo.tsx`     | `/foo`     |
| `src/pages/foo/bar.tsx` | `/foo/bar` |

### Exported Components

Similar to [pages](#pages), exposing components from your Business-Manager module is done by creating a new file in the special `src/exported-components` directory:

```typescript jsx
// src/exported-components/shared-modal.tsx

export default () => {
  return <Modal>...</Modal>;
};
```

### Methods

Exposing methods from your Business-Manager module is done by creating files in the special `src/methods` directory:

```typescript
// src/methods/some-method.ts

export default () => {
  // Do plenty of things
};
```

## Initial Setup

Run `EXPERIMENTAL_FLOW_BM=true npx create-yoshi-app` & choose template `flow-bm`.

This will bootstrap a new Business-Manager application, along with a single (root) page.

## Development

Run `npx yoshi-bm start` to start your Business-Manager application.
This serves your module bundles on http://localhost:3200.

> > In the future, we'll launch your application in production Business-Manager automatically.

Reviewer: ```suggestion
> In the future, your application will be launched in production Business-Manager automatically.
```

## Deployment

Run `npx yoshi-bm build` to build your Business-Manager application.
This builds your module bundle and generates a working `module_MODULE_ID.json` file in your `target` directory, for your [integration with Business-Manager](https://github.com/wix-private/business-manager/blob/master/business-manager-web/docs/module-config-file.md).

## Testing

Run `npx yoshi-bm test` to run your tests with the configured test runner (Jest by default). Pass `--watch` to start it in watch mode.

> ## Customization

Reviewer: I would just change this to `configuration` header instead.

### Module-level Configuration

The following configurations are available by creating a `.module.json` file:

```json
{
  "moduleId": "my-module",
  "routeNamespace": "my-route",
  "topology": {
    "someArtifactsUrl": {
      "artifactId": "com.wixpress.some-artifact"
    }
  }
}
```

---

#### `moduleId`

Use this to override your `moduleId`.

> Defaults to your `artifactId`.

Reviewer: ```suggestion
Defaults to your `artifactId`. (taken from `pom.xml`)
```

#### `routeNamespace`

This prefixes all your pages with the given string.

> Defaults to `''`.

Reviewer: Maybe we can give an example here?

#### `topology`

> Sets your application's topology template. Defaults to:

Reviewer: ```suggestion
Sets your application's topology template. Defaults to your artifactId (taken from `pom.xml`):
```

Author: I wanted to stress that we default to having a single topology variable, it's called `staticsUrl` and it's preconfigured to have your `artifactId`'s url.

But if you supply `topology` in your config - we won't configure `statisUrl`.

Most people eventually need their own artifact's url so I thought it'd be a good default, but I'm not sure saying it defaults to your `artifactId` says all of it.

For example, in Multilingual's case, it looks like this:
https://github.com/wix-private/multilingual/pull/60/files#diff-6969e8fd9380b7c45bf392ee7be4e3c9

(They didn't use `staticsUrl` for their own url, but a different name)

```json
{
  "staticsUrl": {
    "artifactId": "<YOUR_ARTIFACT_ID>"
  }
}
```

### Page-level Configuration

Pages can be customized by adding a `*.json` file with the same name as the appropriate page.
For example, the `src/pages/some-route.tsx` file, will be configured by `src/pages/some-route.json`:

```json
{
  "componentId": "some-component-id",
  "componentName": "some-component-name"
}
```

#### `page.componentId`

Sets the page's `componentId`. Defaults to `<MODULE_ID>.pages.<FILE_NAME>`.

#### `page.componentName`

Sets the page's `componentName`. Defaults to `<MODULE_ID>.pages.<FILE_NAME>`.

### Exported Component-level Configuration

Exported components can be customized by adding a `*.json` file with the same name as the appropriate component file.
For example, the `src/exported-components/some-component.tsx` file, will be configured by `src/exported-components/some-component.json`:

```json
{
  "componentId": "some-component-id"
}
```

#### `exported-component.componentId`

> Sets the component's `componentId`. Defaults to `<MODULE_ID>.componented.<FILE_NAME>`.

Reviewer: `componented`, is it a typo?

Author: Yep :D 
```suggestion
Sets the component's `componentId`. Defaults to `<MODULE_ID>.components.<FILE_NAME>`.
```

### Method-level Configuration

Methods can be customized by adding a `*.json` file with the same name as the appropriate method file.
For example, the `src/methods/some-method.ts` file, will be configured by `src/methods/some-method.json`:

```json
{
  "methodId": "some-method-id"
}
```

> #### `exported-component.componentId`

Reviewer: This looks like a mistake because we're under the Method-level config section

Sets the component's `componentId`. Defaults to `<MODULE_ID>.componented.<FILE_NAME>`.

### Run code in BM's `init()` phase

> Create a file `src/moduleInit.{ts,js}`, for example:

Reviewer: Maybe we can say a few words, that this code will run in each of the pages and it mapped to...
